### PR TITLE
Drivers: U-Blox NEO-M8 Driver implementation + Sample Project

### PR DESCRIPTION
**Pull request by RodinHaker**
Added the driver for the ublox neo-m8 GNSS module.
The data fetching and settings functions can be done using the API. 
A sample project has been added to show the usage of this driver.
As of right now only one message is being parsed and three settings can be done. 
However, the function to write UBX messages to the module is also approachable through the API.

Signed-off-by: Rodin Haker <rodinhaker@hotmail.com>

### CODE REVIEW

File: dts/bindings/gnss/ublox,neom8.yaml
Comment (galak) [inline dts/bindings/gnss/ublox,neom8.yaml:10]: `include i2c-device.yaml`gets you the `reg: required: true` for you.

**Review by galak — CHANGES_REQUESTED** [review]
Remove `reg:...` from dts binding as its handled in i2c-device.yaml

File: include/drivers/gnss/ublox_neo_m8.h
Comment (ycsin) [inline include/drivers/gnss/ublox_neo_m8.h:61]: should use tab

File: include/drivers/gnss/ublox_neo_m8.h
Comment (ycsin) [inline include/drivers/gnss/ublox_neo_m8.h:53]: is '= 0' necessary?

File: include/drivers/gnss/ublox_neo_m8.h
Comment (ycsin) [inline include/drivers/gnss/ublox_neo_m8.h:30]: not sure if it is better to end with comma, same goes for the next enum

File: include/drivers/gnss/ublox_neo_m8.h
Comment (ycsin) [inline include/drivers/gnss/ublox_neo_m8.h:81]: is this millisecond? is uint8_t big enough to fit?

File: samples/drivers/neo_m8/boards/stm32l1_disco.overlay
Comment (ycsin) [inline samples/drivers/neo_m8/boards/stm32l1_disco.overlay:10]: remove?

File: include/drivers/gnss/ublox_neo_m8.h
Comment (ycsin) [inline include/drivers/gnss/ublox_neo_m8.h:84]: seems like this isn't used anywhere?

File: include/drivers/gnss/ublox_neo_m8_defines.h
Comment (ycsin) [inline include/drivers/gnss/ublox_neo_m8_defines.h:1]: Would it be better if you have this file in enum instead of preprocessor macros?

**Review by ycsin — CHANGES_REQUESTED** [review]
clang-format should be ran against these files so that they are compliant with the Zephyr's standard. I think you will also need to propose a generic gnss interface

see #46452, #30564

**Comment by RodinHaker** [comment]
> clang-format should be ran against these files so that they are compliant with the Zephyr's standard. I think you will also need to propose a generic gnss interface
> 
> see #46452, #30564

Should I wait for #46660 to complete and adjust my driver accordingly?

**Comment by sslupsky** [comment]
Had a quick look at the UBX processing and offer the following feedback.  I did something similar when I created a driver for the ublox SAM M8.  As I recall, I found that there are delays in the response from the device.  Moreover, I seem to recall it is possible in some circumstances that you may not actually get a response.  This became problematic for power management purposes because you get stuck waiting for a response in neom8_send_ubx().  The way I solved the issue was to create another task that deals with the response and then notifies the application using a call back mechanism when a response is received.  That way, I could wait on a semaphore which would allow the system to idle and sleep while waiting for a response.

**Comment by ycsin** [comment]
> > clang-format should be ran against these files so that they are compliant with the Zephyr's standard. I think you will also need to propose a generic gnss interface
> > see #46452, #30564
> 
> Should I wait for #46660 to complete and adjust my driver accordingly?

I'd suggest you to wait, and also help shape that API proposal if possible

**Comment by RodinHaker** [comment]
I see that suddenly, after rebasing, this pull request now has many file changes and commits. How do I fix this issue?

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: include/zephyr/drivers/gnss/ublox_neo_m8_defines.h
Comment (RodinHaker) [reply include/zephyr/drivers/gnss/ublox_neo_m8_defines.h:23]: This used to be a file full of defines, however ycsin suggested making this file an enum instead of preprocessed macros.

Comment (carlocaione) [reply include/zephyr/drivers/gnss/ublox_neo_m8_defines.h:23]: right but this is obviously wrong right?

Comment (RodinHaker) [reply include/zephyr/drivers/gnss/ublox_neo_m8_defines.h:23]: These values are for a specific protocol ublox has created called UBX. A ubx frame has a specific structure consisting of 2 header bytes, a class byte, an ID byte, 2 length bytes, a payload and 2 checksum bytes. The UBX_CLASS_NAV is a class with number 0x01. This message would regard the navigational settings. The UBX_ACK_ACK is the value of the payload of an acknowledgement frame. Used to check if the message has been successfully sent and processed. I have included this in the file because it makes it more clear what is happening within the code.

Comment (carlocaione) [reply include/zephyr/drivers/gnss/ublox_neo_m8_defines.h:23]: Are those enums (with the same value) referring to two different things? Then they must go in two different enums.

File: include/zephyr/drivers/gnss/ublox_neo_m8.h
Comment (carlocaione) [reply include/zephyr/drivers/gnss/ublox_neo_m8.h:127]: Given that this PR is not in a review-able state before #46660 is merged, please move this in a draft state.

File: include/zephyr/drivers/gnss/ublox_neo_m8_defines.h
Comment (RodinHaker) [reply include/zephyr/drivers/gnss/ublox_neo_m8_defines.h:23]: Ah okay, I'll change that then

**Comment by sslupsky** [comment]
@RodinHaker I posted my out-of-tree zephyr things here:  https://github.com/sslupsky/zephyr-out-of-tree

There is a driver for the ublox_m8 there if you want to have a look and compare approach.

**Pull request closed without merging**
